src/network/NetworkController.ts

```diff
@@ -78,36 +68,127 @@ export interface NetworkConfig extends BaseConfig {
  * @property provider - RPC URL and network name provider settings
  */
-export interface NetworkState extends BaseState {
+export type NetworkState = {
   network: string;
-  isCustomNetwork: boolean;
+  isCustomNetwork?: boolean;
   provider: ProviderConfig;
-  properties: NetworkProperties;
-}
+  properties?: NetworkProperties;
+};
 
 const LOCALHOST_RPC_URL = 'http://localhost:8545';
 
+const name = 'NetworkController';
+
+export type EthQuery = any;
+
+export type NetworkControllerStateChangeEvent = {
+  type: `NetworkController:stateChange`;
+  payload: [NetworkState, Patch[]];
+};
+
+export type NetworkControllerProviderChangeEvent = {
+  type: `NetworkController:providerChange`;
+  payload: [ProviderConfig];
+};
+
+export type NetworkControllerGetProviderConfig = {
+  type: `NetworkController:getProviderConfig`;
+  handler: () => ProviderConfig;
+};
+
+export type NetworkControllerGetEthQuery = {
+  type: `NetworkController:getEthQuery`;
+  handler: () => EthQuery;
+};
+
+export type NetworkControllerMessenger = RestrictedControllerMessenger<
+  typeof name,
+  NetworkControllerGetProviderConfig | NetworkControllerGetEthQuery,
+  NetworkControllerStateChangeEvent | NetworkControllerProviderChangeEvent,
+  string,
+  string
+>;
+
+export type NetworkControllerOptions = {
+  messenger: NetworkControllerMessenger;
+  infuraProjectId?: string;
+  state?: Partial<NetworkState>;
+};
+
+const defaultState: NetworkState = {
+  network: 'loading',
+  isCustomNetwork: false,
+  provider: { type: MAINNET, chainId: NetworksChainId.mainnet },
+  properties: { isEIP1559Compatible: false },
+};
+
 /**
  * Controller that creates and manages an Ethereum network provider
  */
 export class NetworkController extends BaseController<
-  NetworkConfig,
-  NetworkState
+  typeof name,
+  NetworkState,
+  NetworkControllerMessenger
 > {
-  private ethQuery: any;
+  private ethQuery: EthQuery;
 
   private internalProviderConfig: ProviderConfig = {} as ProviderConfig;
 
+  private _infuraProjectId: string | undefined;
+
   private mutex = new Mutex();
 
+  constructor({ messenger, state, infuraProjectId }: NetworkControllerOptions) {
+    super({
+      name,
+      metadata: {
+        network: {
+          persist: true,
+          anonymous: false,
+        },
+        isCustomNetwork: {
+          persist: true,
+          anonymous: false,
+        },
+        properties: {
+          persist: true,
+          anonymous: false,
+        },
+        provider: {
+          persist: true,
+          anonymous: false,
+        },
+      },
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+    this._infuraProjectId = infuraProjectId;
+    this.messagingSystem.registerActionHandler(
+      `${this.name}:getProviderConfig`,
+      () => {
+        return this.state.provider;
+      },
+    );
+
+    this.messagingSystem.registerActionHandler(
+      `${this.name}:getEthQuery`,
```

Comment:
Interesting - why expose an `EthQuery` instance rather than the provider itself?

While the current provider is _not great_, we were planning on replacing it with a new EIP-1193 provider that we could call `request` on directly. And at that point, we could drop `EthQuery`.

Reply:
might be good to expose that one as well, though there seems to be more examples where either 
1. ethQuery is pulled off the networkController instance 
2. a new ethQuery is constructed from the result of subbing to network controller state changes

I'd like to avoid optimizing around the use of ethQuery vs. provider stuff for this PR if possible. We absolutely should move towards 1193 provider though in the very near future.

Reply:
I wasn't able to find any examples of where the networkController ethQuery is directly used. I'm just seeing the second example, where it's re-constructed upon each network change.

We have to re-construct it upon network change because the provider given is not updated for the new network. This is why the extension doesn't expose the provider directly, it exposes a provider proxy. If we were to do the same thing here, there would be no need to re-construct EthQuery either. This is already one of the steps we'll need to take to bring the two controllers to feature parity, so that seems like a better way of eliminating the extra work in that second example.

We'll need to expose the provider directly anyway because it's used directly in mobile and extension for the RPC pipeline.

Reply:
So are we saying that this is a fine change for now or that we need something different?

Reply:
proposed way forward:
1. Ill add an action handler for getProvider. 
2. In a future PR, we should remove getEthQuery, and replace use of ethQuery with calls directly to the 1193 provider.

Reply:
3. in future PR, change `providerChange` event to `providerConfigChange`, and change `state.provider` to `state.providerConfig`
4. simplify the logic flow, make it async where required, update tests to properly await networkcontroller initialization, make the underlying provider easily mockable for testing (ideally not having to use nock)

Reply:
sorry for using this thread as a notepad ealier... however, i have a commit that I can push to this branch that (I think) addresses the above concerns:
https://github.com/MetaMask/controllers/compare/refactor-network-controller-v2...refactor-network-controller-v2-oct3?expand=1
let me know what you think